server/src/test/java/org/elasticsearch/action/bulk/TransportShardBulkActionTests.java

```diff
@@ -1072,6 +1091,70 @@ public void testPerformOnPrimaryReportsBulkStats() throws Exception {
         latch.await();
     }
 
+    public void testNoopMappingUpdateInfiniteLoopPrevention() throws Exception {
```

Comment:
Can we add a test that verifies that a noop retry succeeds?

Reply:
I've added a unit test. In addition to that, `DynamicMappingIT#testConcurrentDynamicUpdates` also tests this.